src/section_list_cache.rs

```diff
@@ -0,0 +1,159 @@
+            for list in lists_to_remove {
+                let _ = map.remove(&list);
+            }
+            if map.is_empty() {
+                to_remove.push(*prefix);
+            }
+        }
+        // prune prefixes with no group lists
+        for prefix in to_remove {
+            let _ = self.signatures.remove(&prefix);
+        }
+
+        let to_remove = self.signed_by
+            .iter()
+            .filter(|&(_, map)| map.is_empty())
+            .map(|(pub_id, _)| *pub_id)
+            .collect_vec();
+        // prune pub_ids signing nothing
+        for pub_id in to_remove {
+            let _ = self.signed_by.remove(&pub_id);
+        }
+    }
+
+    fn update_lists_cache(&mut self, our_section_size: usize) {
+        for (prefix, map) in &self.signatures {
+            // find the entries with the most signatures
+            let entries = map.iter()
+                .map(|(list, sigs)| (list, sigs.len()))
+                .sorted_by(|lhs, rhs| rhs.1.cmp(&lhs.1));
+            if let Some(&entry) = entries.first() {
+                // entry.0 = list, entry.1 = num of signatures
+                if 100 * entry.1 >= QUORUM * our_section_size {
+                    // we have a list with a quorum of signatures
+                    let signatures = unwrap!(map.get(entry.0));
+                    let _ = self.lists_cache.insert(*prefix, (entry.0.clone(), signatures.clone()));
+                }
+            }
+        }
+    }
+
+    fn remove_signatures_for_prefix_by(&mut self, prefix: Prefix<XorName>, author: PublicId) {
+        // vector of tuples (prefix, group list) to be removed
+        let to_remove = self.signed_by
+            .get(&author)
+            .map_or(Default::default(), |map| {
+                map.iter()
+                    .filter(|&(p, _)| p.is_compatible(&prefix))
+                    .map(|(&prefix, list)| (prefix, list.clone()))
+                    .collect_vec()
+            });
+        for (prefix, list) in to_remove {
+            // remove the signatures from self.signatures
+            let _ = self.signatures.get_mut(&prefix).map_or(None, |map| {
+                map.get_mut(&list).map_or(None, |sigmap| sigmap.remove(&author))
+            });
+            // remove those entries from self.signed_by
+            let _ = self.signed_by.get_mut(&author).map_or(None, |map| map.remove(&prefix));
+        }
+
+        self.prune();
+    }
+
+    /// Removes all signatures authored by `author`
+    pub fn remove_signatures_by(&mut self, author: PublicId, our_section_size: usize) {
+        if let Some(lists) = self.signed_by.remove(&author) {
+            for (prefix, list) in lists {
+                let _ = self.signatures.get_mut(&prefix).map_or(None, |map| {
+                    map.get_mut(&list).map_or(None, |sigmap| sigmap.remove(&author))
+                });
+            }
+            self.prune();
+            self.update_lists_cache(our_section_size);
+        }
+    }
```

Comment:
As discussed, let's stick to this approach for now; the signatures are preserved in the cache anyway.